dvc/remote/base.py

```diff
@@ -849,44 +702,243 @@ def exists_with_progress(path_info):
+        # for performance overhead from large lists/sets.
+        # From testing with S3, for remotes with 1M+ files, object_exists is
+        # faster until len(checksums) is at least 10k~100k
+        if remote_size > self.tree.TRAVERSE_THRESHOLD_SIZE:
+            traverse_weight = (
+                traverse_pages * self.tree.TRAVERSE_WEIGHT_MULTIPLIER
+            )
+        else:
+            traverse_weight = traverse_pages
+        if len(checksums) < traverse_weight:
+            logger.debug(
+                "Large remote ('{}' checksums < '{}' traverse weight), "
+                "using object_exists for remaining checksums".format(
+                    len(checksums), traverse_weight
+                )
+            )
+            return (
+                list(indexed_checksums)
+                + list(checksums & remote_checksums)
+                + self.tree.list_checksums_exists(
+                    checksums - remote_checksums, jobs, name
+                )
+            )
+
+        logger.debug(
+            "Querying '{}' checksums via traverse".format(len(checksums))
+        )
+        remote_checksums = set(
+            self.tree.list_checksums_traverse(
+                remote_size, remote_checksums, jobs, name
+            )
+        )
+        return list(indexed_checksums) + list(
+            checksums & set(remote_checksums)
+        )
+
+    @classmethod
     @index_locked
-    def gc(self, named_cache, jobs=None):
+    def gc(cls, named_cache, remote, jobs=None):
+        tree = remote.tree
         used = set(named_cache.scheme_keys("local"))
 
-        if self.scheme != "":
-            used.update(named_cache.scheme_keys(self.scheme))
+        if tree.scheme != "":
+            used.update(named_cache.scheme_keys(tree.scheme))
 
         removed = False
         # checksums must be sorted to ensure we always remove .dir files first
         for checksum in sorted(
-            self.all(jobs, str(self.path_info)),
-            key=self.is_dir_checksum,
+            tree.all(jobs, str(tree.path_info)),
+            key=tree.is_dir_checksum,
             reverse=True,
         ):
             if checksum in used:
                 continue
-            path_info = self.checksum_to_path_info(checksum)
-            if self.is_dir_checksum(checksum):
+            path_info = tree.checksum_to_path_info(checksum)
+            if tree.is_dir_checksum(checksum):
                 # backward compatibility
-                self._remove_unpacked_dir(checksum)
-            self.tree.remove(path_info)
+                tree._remove_unpacked_dir(checksum)
+            tree.remove(path_info)
             removed = True
-        if removed:
-            self.index.clear()
+
+        if removed and hasattr(remote, "index"):
+            remote.index.clear()
         return removed
 
-    def _remove_unpacked_dir(self, checksum):
-        pass
 
+class CloudCache:
```

Comment:
Quite strange why it is CloudCache and not just Cache? And why LocalCache is the only exception?

Reply:
Ah, I guess because of the push/pull. Ok, makes sense.

Reply:
I went with CloudCache to avoid the collision with `dvc.cache.Cache`, but maybe this should be Cache, and the other one should be `RepoCache` or something along those lines.

Reply:
@pmrowla Makes sense. I'm fine with CloudCache for now.

Reply:
We can also get rid of LocalCache in the future by making push/pull/status more generic for syncing data between any Cache and any Remote, rather than having it be specific to LocalCache, but I don't think there's any pressing need for that right now.